go/tools/builders/asm.go

```diff
@@ -137,12 +141,34 @@ func buildSymabisFile(goenv *env, sFiles, hFiles []fileInfo, asmhdr string) (str
 	return symabisName, err
 }
 
-func asmFile(goenv *env, srcPath string, asmFlags []string, outPath string) error {
+func asmFile(goenv *env, srcPath, packagePath string, asmFlags []string, outPath string) error {
 	args := goenv.goTool("asm")
 	args = append(args, asmFlags...)
+	// The package path has to be specified as of Go 1.19 or the resulting
+	// object will be unlinkable, but the -p flag is also only available
+	// since Go 1.19.
+	if packagePath != "" && isGo119OrHigher() {
+		args = append(args, "-p", packagePath)
+	}
 	args = append(args, "-trimpath", ".")
 	args = append(args, "-o", outPath)
 	args = append(args, "--", srcPath)
 	absArgs(args, []string{"-I", "-o", "-trimpath"})
 	return goenv.runCommand(args)
 }
+
+var goMinorVersionRegexp = regexp.MustCompile(`^go1\.(\d+)`)
+
+func isGo119OrHigher() bool {
+	match := goMinorVersionRegexp.FindStringSubmatch(runtime.Version())
+	if match == nil {
+		// Developer version or something with an unparseable version string,
+		// assume Go 1.19 or higher.
+		return true
+	}
+	minorVersion, err := strconv.Atoi(match[1])
+	if err != nil {
+		return true
+	}
+	return minorVersion >= 19
+}
```

Comment:
A quick test on this approach for those who wonder "how would this work with patch versions"

```
~/test> cat main.go
package main

import (
    "fmt"
    "regexp"
    "runtime"
)

func main() {
    fmt.Println("version:", runtime.Version())

    goMinorVersionRegexp := regexp.MustCompile(`^go1\.(\d+)`)
    matches := goMinorVersionRegexp.FindStringSubmatch(runtime.Version())
    for i, m := range matches {
        fmt.Printf("match no %d: %s\n", i, m)
    }
}
~/test> go run main.go
version: go1.18.4
match no 0: go1.18
match no 1: 18
```

So it works :)

Reply:
Yeah, that's why I went with a regexp instead of simple split logic - I was worried of patch and rc/beta versions.